Add htaccess to redirect if not found to index

the-hotmann opened the pull request:
If (file not found or directory not found) 
then > redirect to index.html

Related to https://github.com/shlinkio/shlink-web-client/issues/196

the-hotmann committed:
Add htaccess to redirect if not found to index

If file not fount or directory not found redirect to index.html

acelaya commented:
Awesome! Thanks a lot @MartinH0 😁

Just have two comments.

* The htaccess file should be located inside the `public` directory, otherwise the build process won't include it in the final dist file.
* This rule will have a side effect. When accidentally introducing a typo on an asset route (like an image), Apache will serve the index file instead, making it hard to detect. Ideally, when the path has an extension, Apache should continue returning a 404, and only for paths without extension (which are usually handled client-side) it should fall back to the index file.

For the second point, you can take a look at how it's done for the nginx inside the docker image https://github.com/shlinkio/shlink-web-client/blob/master/config/docker/nginx.conf

the-hotmann committed:
Deleted .htaccess at wrong directory

Sorry fucked it up, will correct it.

the-hotmann committed:
Add htaccess to redirect if not found to index

If (file not found or directory not found)
then > redirect to index.html

the-hotmann commented:
Wait. I will optimize it. So just files without file extensions will be redirected

the-hotmann committed:
Updated htaccess to meet required functions.

If a file gets called it will be redirected to index.html

But not if it the requested File does contain a dot (and with this does have a file extension.

If you call:
links.domain.de/notexistingfile.jpg 
It will trigger 404

If you call:
links.domain.de/server/[CODE-CODE-CODE]/list-short-urls/1
It will redirect the call to index.html

the-hotmann commented:
Now you can merge if you like it.

acelaya commented:
Thanks for the contribution! I will test it later and merge it if everything goes as expected.

the-hotmann committed:
Updated htaccess

the-hotmann committed:
Updates htaccess

return 404 error if static assets does not exist

the-hotmann committed:
updated htaccess

deleted second json (just needed once)

the-hotmann committed:
Update .htaccess

1. added more comments.
2. added NC Tag for making all the static assets case insensetive ("jpg" now matches "jpg" and "JPG" and so on)
3. transformed "jpe|jpeg" into "jpe?g" as its regex for the same, but shorter
4. changed line 8 from "+" to "*" to match everything, also zero times the wildcard

the-hotmann committed:
Update .htaccess

1. removed $ (dollar sign from line 14
2. changed line 8 from ".*" to "(.*)"

the-hotmann committed:
Updated nginx.conf (optimization for future)

1. changed location from "~" (case sensitive!) to "~*" (case insensitive!) to also match uppercase static assets. (http://nginx.org/en/docs/http/ngx_http_core_module.html#location)
2. added regex "jpe?g" to match "jpg" and "jpeg" in one command.

acelaya commented on config/docker/nginx.conf:
```diff
@@ -5,7 +5,7 @@ server {
 
     # When requesting static paths with extension, try them, and return a 404 if not found
-    location ~ .+\.(css|js|html|png|jpg|jpeg|gif|bmp|ico|json|csv|otf|eot|svg|svgz|ttf|woff|woff2|ijmap|pdf|tif|map) {
+    location ~* .+\.(css|js|html|png|jpe?g|gif|bmp|ico|json|csv|otf|eot|svg|svgz|ttf|woff|woff2|ijmap|pdf|tif|map) {
```
What does this extra `*` do? What's the difference between `location ~*` and `location ~`?

acelaya replied:
Never mind. Just found the explanation here https://www.digitalocean.com/community/tutorials/understanding-nginx-server-and-location-block-selection-algorithms :sweat_smile: 

Good catch.

acelaya reviewed (approved):
Just tested all changes, and they work as expected. Good job :)
